src/Storages/MergeTree/MergeTreeIndexReadResultPool.cpp

```diff
@@ -94,13 +101,346 @@ SkipIndexReadResultPtr MergeTreeSkipIndexReader::read(const RangesInDataPart & p
+{
+    size_t old_size = filter.size();
+    filter.resize_fill(old_size + num_rows);
+    UInt8 * pos = &filter[old_size];
+    size_t ending_row = starting_row + num_rows;
+
+    if (type == BitmapType::Bitmap32)
+    {
+        roaring::api::roaring_uint32_iterator_t it;
+        roaring_iterator_init(data.bitmap32, &it);
+        if (!roaring_uint32_iterator_move_equalorlarger(&it, starting_row))
+            return false;
+
+        bool has_value = false;
+        while (it.current_value < ending_row)
+        {
+            has_value = true;
+            pos[it.current_value - starting_row] = 1;
+            if (!roaring_uint32_iterator_advance(&it))
+                break;
+        }
+        return has_value;
+    }
+    else
+    {
+        chassert(type == BitmapType::Bitmap64);
+        /// NOTE: roaring64 requires a heap-allocated opaque iterator (unlike 32-bit)
+        auto * it = roaring::api::roaring64_iterator_create(data.bitmap64);
+        /// There is no way to recover a failed allocation inside roaring64.
+        /// See https://github.com/RoaringBitmap/CRoaring/issues/638
+        if (it == nullptr)
+            throw Exception(ErrorCodes::MEMORY_LIMIT_EXCEEDED, "Failed to allocate roaring64 iterator");
+
+        if (!roaring::api::roaring64_iterator_move_equalorlarger(it, starting_row))
+        {
+            roaring::api::roaring64_iterator_free(it);
+            return false;
+        }
+
+        auto val = roaring64_iterator_value(it);
+        bool has_value = false;
+        while (val < ending_row)
+        {
+            has_value = true;
+            pos[val - starting_row] = 1;
+            if (!roaring::api::roaring64_iterator_advance(it))
+                break;
+            val = roaring64_iterator_value(it);
+        }
+        roaring::api::roaring64_iterator_free(it);
+        return has_value;
+    }
+}
+
+template ProjectionIndexBitmapPtr ProjectionIndexBitmap::createFromRange<UInt32>(UInt32 start, UInt32 end);
+template ProjectionIndexBitmapPtr ProjectionIndexBitmap::createFromRange<UInt64>(UInt64 start, UInt64 end);
+
+template bool ProjectionIndexBitmap::contains<UInt32>(UInt32 value);
+template bool ProjectionIndexBitmap::contains<UInt64>(UInt64 value);
+
+template void ProjectionIndexBitmap::add<UInt32>(UInt32 value);
+template void ProjectionIndexBitmap::add<UInt64>(UInt64 value);
+
+SingleProjectionIndexReader::SingleProjectionIndexReader(
+    std::shared_ptr<MergeTreeReadPoolProjectionIndex> pool,
+    PrewhereInfoPtr prewhere_info,
+    const ExpressionActionsSettings & actions_settings,
+    const MergeTreeReaderSettings & reader_settings)
+    : projection_index_read_pool(std::move(pool))
+    , processor(std::make_unique<MergeTreeSelectProcessor>(
+          std::static_pointer_cast<IMergeTreeReadPool>(projection_index_read_pool),
+          std::make_unique<MergeTreeProjectionIndexSelectAlgorithm>(),
```

Comment:
Do we really need to create a pool and `MergeTreeSelectProcessor`? Is it required for the correct calculation of PREWHERE? That's ok, but looks a bit overcomplicated.

Maybe we can just create a `MergeTreeReadTask` and call `MergeTreeReadTask::read`.